### ISSUE TITLE
MediaUploadCheck component should be usable outside gutenberg editor

### ISSUE THREAD

**Issue by juanRabaa** [WordPress/gutenberg#40698]
## What problem does this address?
The `MediaUpload` component that comes from the `@wordpress/block-editor` package needs to be wrapped around the `MediaUploadCheck` component to make sure that the current user has media upload permissions (as stated in https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/media-upload/README.md).

To verify this permissions, the `MediaUploadCheck` makes use of the `core/block-editor` store, checking the setting `mediaUpload`. Because of this, the component is not usable outside of a gutenberg editor context.

## What is your proposed solution?

My understanding is that `MediaUploadCheck` shouldn't depend on the editor, as the user capabilities to `read` or `update` media is something that can be checked outside of it.

The verification should be done using the `canUser` method from the `core` store instead of the `core/block-editor`.

I ended up using something like this in my code. This also displays an spinner if the capability check hasn't yet ended.

```JS
import React from "react";
const { useSelect } = wp.data;
const { Spinner } = wp.components;

export default function MediaUploadCheck(props) {
    const {
        fallback = null,
        children,
    } = props;

    const {
        checkingPermissions,
        hasUploadPermissions,
    } = useSelect(select => {
        const core = select("core");
        return {
            hasUploadPermissions: core.canUser("read", "media"),
            checkingPermissions: !core.hasFinishedResolution("canUser", [ "read", "media"]),
        };
    });

    return <>
        {checkingPermissions && <Spinner/>}
        {(!checkingPermissions && hasUploadPermissions) ? children : fallback}
    </>;
}
```

**Comment by juanRabaa** [comment]
I pressed enter unintentionally and posted the issue before filling the data :(. Will reopen when edited

**Comment by juanRabaa** [comment]
Completed the issue

**Comment by talldan** [comment]
I think it has been implemented this way (by checking `mediaUpload` that's defined in the block editor settings) so that non-WordPress block editor implementations can support media uploading.

It is interesting the way this has been done. The `MediaUpload` component doesn't actually use the `mediaUpload` function, so I'm not sure why it's needed. `MediaUpload` is a slot, and checking permissions could be handled by the code that renders within that slot rather than code that renders outside the slot.

Having said all that, It might too late to change that now, as it would potentially break backwards compatibility.

**Comment by juanRabaa** [comment]
@talldan yes, the `MediaUpload` component works outside of the editor, but in the only documentation that there is about this block its recommended to be used inside the `MediaUploadCheck` component (see [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/media-upload/README.md) )

Yes you could actually do like I did your own implementation of a component that checks for the permissions, but I think the one that already exists (`MediaUploadCheck`) doesnt have to be dependent on the editor. The logic behind the verification of capabilities is made in the component itself, so changing it wouldn't break backwards compatibility.

I may be missing something about the way it is implemented right now. Maybe there is a reason behind checking `getSettings().mediaUpload` using the `core/block-editor`.

**Comment by talldan** [comment]
> I may be missing something about the way it is implemented right now. Maybe there is a reason behind checking getSettings().mediaUpload using the core/block-editor.

Yes, it's the reason I said above 😄 

The `block-editor` package can be used in non-WordPress implementations. For example another CMS or another application entirely can install `@wordpress/block-editor` from npm and implement their own block editor. When doing so, they can specify their own media uploading functionality by implementing a `mediaUpload` function in the editor settings.

`canUser` couldn't be used in place of `getSettings().mediaUpload` because it checks WordPress privileges using the REST API.

`MediaUploadCheck` could be implemented the way you desire, but it would need to be in a different package. For example, the `wordpress/media-utils` package could have a version that works that way.
